go/nbs/table_set.go

```diff
@@ -81,27 +85,141 @@ type tablePersister interface {
 }
 
 type s3TablePersister struct {
-	s3     s3svc
-	bucket string
+	s3       s3svc
+	bucket   string
+	partSize int
+}
+
+func (s3p s3TablePersister) Open(name addr, chunkCount uint32) chunkSource {
+	return newS3TableReader(s3p.s3, s3p.bucket, name, chunkCount)
+}
+
+type s3UploadedPart struct {
+	idx  int64
+	etag string
 }
 
 func (s3p s3TablePersister) Compact(mt *memTable, haver chunkReader) (name addr, chunkCount uint32) {
 	name, data, chunkCount := mt.write(haver)
 
 	if chunkCount > 0 {
-		_, err := s3p.s3.PutObject(&s3.PutObjectInput{
-			Bucket:        aws.String(s3p.bucket),
-			Key:           aws.String(name.String()),
-			Body:          bytes.NewReader(data),
-			ContentLength: aws.Int64(int64(len(data))),
+		result, err := s3p.s3.CreateMultipartUpload(&s3.CreateMultipartUploadInput{
+			Bucket: aws.String(s3p.bucket),
+			Key:    aws.String(name.String()),
 		})
 		d.PanicIfError(err)
+		uploadID := *result.UploadId
+
+		multipartUpload, err := s3p.uploadParts(data, name.String(), uploadID)
+		if err != nil {
+			_, abrtErr := s3p.s3.AbortMultipartUpload(&s3.AbortMultipartUploadInput{
+				Bucket:   aws.String(s3p.bucket),
+				Key:      aws.String(name.String()),
+				UploadId: aws.String(uploadID),
+			})
+			d.Chk.NoError(abrtErr)
+			panic(err) // TODO: Better error handling here
+		}
+
+		_, err = s3p.s3.CompleteMultipartUpload(&s3.CompleteMultipartUploadInput{
+			Bucket:          aws.String(s3p.bucket),
+			Key:             aws.String(name.String()),
+			MultipartUpload: multipartUpload,
+			UploadId:        aws.String(uploadID),
+		})
+		d.Chk.NoError(err)
 	}
 	return name, chunkCount
 }
 
-func (s3p s3TablePersister) Open(name addr, chunkCount uint32) chunkSource {
-	return newS3TableReader(s3p.s3, s3p.bucket, name, chunkCount)
+func (s3p s3TablePersister) uploadParts(data []byte, key, uploadID string) (*s3.CompletedMultipartUpload, error) {
+	sent, failed, done := make(chan s3UploadedPart), make(chan error), make(chan struct{})
+
+	numParts := getNumParts(len(data), s3p.partSize)
+	var wg sync.WaitGroup
+	wg.Add(numParts)
+	sendPart := func(partNum int) {
+		defer wg.Done()
+
+		start, end := (partNum-1)*s3p.partSize, partNum*s3p.partSize
+		if partNum == numParts { // If this is the last part, make sure it includes any overflow
+			end = len(data)
+		}
+		result, err := s3p.s3.UploadPart(&s3.UploadPartInput{
+			Bucket:     aws.String(s3p.bucket),
+			Key:        aws.String(key),
+			PartNumber: aws.Int64(int64(partNum)),
+			UploadId:   aws.String(uploadID),
+			Body:       bytes.NewReader(data[start:end]),
+		})
+		if err != nil {
+			failed <- err
+			return
+		}
+
+		select {
+		case sent <- s3UploadedPart{int64(partNum), *result.ETag}:
+		case <-done:
```

Comment:
is the point of reading from `done` here just to ensure that this goroutine exits (and doesn't sitting blocking on the send to `sent` forever?